Add Best Constrained Path Menu

MarvinTorres opened the pull request:
This helps resolve #22

Demo of current iteration:
![Peek 2020-07-28 21-17](https://user-images.githubusercontent.com/26232289/88745440-116d0880-d118-11ea-972b-fc1a649350d5.gif)

This updates the UI to support searches for multiple shortest constrained flexible paths. This change affects only the interface - the logic will be the subject of a separate pull request.

Elements Added:
- A dropdown menu for "_Best Constrained Paths_"
   - Options to filter out paths with edges that fail to meet the following requirements:
      - Bandwidth less than minimum
      - Reliability less than minimum
      - Delay more than maximum
      - Utilization more than maximum
      - Priority less than minimum
      - Owner different than the requested owner
   - Option to set the flexibility of the filtering.
      - 0 (none): include only edges that hit all requirements.
      - 0 < n < 6 (partial): include only edges that have at most n misses (n is an integer). 
      - 6 (full): allow all edges regardless of misses.
- A scrollbar in case the elements overflow.

Screenshots:
Open
![Screenshot from 2020-06-10 16-54-00](https://user-images.githubusercontent.com/26232289/84318005-bd906b00-ab3b-11ea-9f1d-daab4719287c.png)
Closed
![Screenshot from 2020-06-10 16-55-37](https://user-images.githubusercontent.com/26232289/84317987-b701f380-ab3b-11ea-8a8d-8891a6dcb0c0.png)

Ktmi committed:
Finished webui
 + Adds in support for constraints
 + Adds in support for flexibility

Ktmi committed:
Removed code from web ui used to help debug

MarvinTorres committed:
updated change log

MarvinTorres committed:
set metric attribute input values to 0

MarvinTorres committed:
removed the icons next to the sliders

MarvinTorres committed:
declared var inside of loop block

MarvinTorres committed:
compacted sliders and added scrollbar

MarvinTorres commented:
I am changing the UI design to support picking flexible and inflexible metrics.

MarvinTorres committed:
Renamed variable flexible to depth

MarvinTorres commented:
I did not have enough time to work on the UI too much - it is not much different than before. I would like another review if possible.

cmagnobarbosa commented:
> I did not have enough time to work on the UI too much - it is not much different than before. I would like another review if possible.

Sure, I will review.

MarvinTorres committed:
Added accordion item for new best constrained path

cmagnobarbosa commented:
Hello @MarvinTorres, exist a small problem when the browser is resized. The map doesn’t adjust creating a blank bar below the map.

![Captura de tela de 2020-06-24 18-41-07](https://user-images.githubusercontent.com/10928556/85630623-5cdc5480-b64a-11ea-8558-de1e84db307f.png)

A first, solution to this problem is to disable the scrollbar on the map. But this solution reduces the responsive.

Solution:
Add the code below in `pathfinder/ui/k-toollbar/main.kytos`
``` html
<style type="text/css">
body{
    overflow-y:hidden;
}
</style>
```
Result:
![first_result](https://user-images.githubusercontent.com/10928556/85631980-daa15f80-b64c-11ea-8f39-00f1f4721d1d.png)

I am looking for a better solution.

MarvinTorres committed:
Continued working on new component

MarvinTorres commented:
Thank you @cmagnobarbosa, I will implement your changes. EDIT: It worked!

I managed to add a new component that supports searching for flexible and inflexible metrics. Below is a W.I.P screenshot of its current iteration. Feel free to comment on it - its design is tentative.

![updated](https://user-images.githubusercontent.com/26232289/85639849-0a099980-b658-11ea-9fc0-6fcbd9047971.png)

MarvinTorres committed:
Implemented cmagnobarbosa's map resize fix

MarvinTorres committed:
Fixed logic error in get_constrained_paths_v4

 - self.v4_is_flexible[checked] was comparing strings, not boolean.
  - Will need to change code to use bools

MarvinTorres committed:
Updated get_constrained_paths_v4 to reuse bools.

MarvinTorres committed:
Updated API endpoints and removed unnecessary menu
 -  "Best Constrained Path With Flexible Metrics" now "Best Constrained Path"
 -  Original "Best Constrained Path" menu removed since it was redundant
 - v3 endpoint removed
 - v4 endpoint changed to v2/best-constrained-paths
 - variable names changed to reflect removed menu

MarvinTorres commented:
Hello everyone, I updated the UI once more. To demonstrate this, I added a GIF that shows the updated UI in the original post.

rmotitsuki commented on ui/k-toolbar/main.kytos:
```diff
@@ -1,20 +1,92 @@
-    <k-accordion>
-      <k-accordion-item title="Best Path">
-        <k-dropdown icon="circle-o" title="Source:" :options="get_interfaces"
+    <div style="overflow-y: auto; height:550px;">
```
The height property can be more responsive using:
height: calc(100vh - 60px);
It is not the best solution, but it will solve the blank lines below the menu, and the height will be dynamic

MarvinTorres replied:
I added your changes in the latest push. Thank you!

rmotitsuki commented on ui/k-toolbar/main.kytos:
```diff
@@ -1,20 +1,92 @@
-    <k-accordion>
-      <k-accordion-item title="Best Path">
-        <k-dropdown icon="circle-o" title="Source:" :options="get_interfaces"
+    <div style="overflow-y: auto; height:550px;">
```
It is not recommend to use the inline style attribute for HTML elements.
It would be better to move all style attributes in this file to the style tag.

MarvinTorres replied:
Fixed. Thank you!

rmotitsuki commented:
Kytos also has a "compact view" for the menu. It changes the view when you click on 'Kytos' in the upper left corner.
I think that it will not fit all the options, so you can put just the main filters in there.
![image](https://user-images.githubusercontent.com/10867136/86642407-402d0e80-bfb2-11ea-8991-7bc170d4a502.png)

MarvinTorres committed:
Changed inline css tags to internal css tags

MarvinTorres commented:
@rmotitsuki I would like to add a compact version of this menu. But I was unable to find documentation on how to do so. I could be missing something, though.

I could implement a simplified menu, but I would like to design it such that it does not show up in the uncompacted version. But I have no idea where to begin implementing that design. If you would help me, I would appreciate it.

rmotitsuki commented:
> @rmotitsuki I would like to add a compact version of this menu. But I was unable to find documentation on how to do so. I could be missing something, though.
> 
> I could implement a simplified menu, but I would like to design it such that it does not show up in the uncompacted version. But I have no idea where to begin implementing that design. If you would help me, I would appreciate it.

@MarvinTorres 
The compact version is mentioned in the [Web Interface](https://docs.kytos.io/kytos/developer/web-ui/?highlight=compacted) documentation, but I was unable to find an implementation document.
I will write a document but it may take a while. 
You can check the source code of the UI component  to understand how it works.
The [Textarea ](https://github.com/kytos/ui/blob/743c1ddacd6e67fa025fd5999c8c6411e1a82c5e/src/components/kytos/inputs/Textarea.vue) component does not appear in the compacted format. It uses a class _no-compact_ in the HTML template to hide it.
The [Input ](https://github.com/kytos/ui/blob/4ff56ae46343e43a11071d1133543d4f898334de/src/components/kytos/inputs/Input.vue) component is displayed in the compacted format, so there is a _compacted_ class in the style tag.

MarvinTorres commented:
@rmotitsuki Thank you for your help! I will look up the source files as you suggested.

If I have any more questions I will let you know.

gleybersonandrade commented:
Hello @MarvinTorres, I was testing your user interface and it's great, thanks bro! I have a question: If I selected an Undesired Link, can I deselect it?

![image](https://user-images.githubusercontent.com/20271875/87673100-583b2580-c74a-11ea-9392-4bc530410625.png)

MarvinTorres commented:
@gleybersonandrade I messed with the UI a bit and found that I could unselect an option by clicking on it while holding `ctrl`.

gleybersonandrade commented:
@MarvinTorres sure, so simple! :) Thanks man

MarvinTorres commented:
@gleybersonandrade @rmotitsuki @josemauro @cmagnobarbosa I would like to merge the *Best Path* and *Best Constrained Path* accordion items into one menu. Let me know what you think of that idea.

MarvinTorres committed:
Multiple changes to improve UX
   - Added "Minimum Flexible Hits" text to menu
   - Changed "maximum_misses" var name to "minimum_flexible_hits"

MarvinTorres committed:
Removed changelog update
